Add script for crate structure comparison

igorostrowskiq opened the pull request:
Add script for crate structure analysis and comparison with older results

arkjedrz commented on scripts/coverage_checker.py:
```diff
@@ -0,0 +1,346 @@
+import argparse
```
Missing shebang and docstring.

igorostrowskiq replied:
Done

arkjedrz commented on scripts/coverage_checker.py:
- File is missing exe rights.
- Please take `ruff` rules from #62 and run `ruff` on this file. I get some issues reported.

arkjedrz commented on scripts/coverage_checker.py:
```diff
@@ -0,0 +1,346 @@
+
+
+class TreeNode:
+    def __init__(self, entries: list[Entry], ndx: Indexer, parent: Union["TreeNode", None] = None):
```
`Union["TreeNode", None]` -> `"TreeNode" | None`

igorostrowskiq replied:
Used `Optional`

arkjedrz reviewed (changes requested):
- I don't think parsing languages using scripts is a good idea. In my experience - it is working shabby at best, and causing unsustainable burden to maintain at worst.
- I feel there might be an overlap with what is expected with https://github.com/cargo-public-api/cargo-public-api
- I'm missing any kind of rationale or example what's this script for.
- I was unable to find one file from Persistency repo that will work with this scripts.

igorostrowskiq commented on scripts/coverage_checker.py:
```diff
@@ -0,0 +1,346 @@
+        if found != -1:
+            return Entry(line, item_type, found)
+
+    raise RuntimeError("No known item type found for parsed line: " + line)
```
Done

igorostrowskiq commented on scripts/coverage_checker.py:
```diff
@@ -0,0 +1,346 @@
+
+        self.data = line_parts[0]
+        self.comment = line_parts[-1] if len(line_parts) > 1 else None
+        self.tab_count = line.count("\t") if self.comment is not None else None
```
Done

igorostrowskiq commented on scripts/coverage_checker.py:
```diff
@@ -0,0 +1,346 @@
+        tokens = data_line.split()
+
+        if tokens[0] == "crate":
+            # Crate has no visibility specifier, set pub for convinience
```
Fixed

igorostrowskiq commented on scripts/coverage_checker.py:
```diff
@@ -0,0 +1,346 @@
+        for ndx, token in enumerate(tokens):
+            token = token.strip()
+            if token.endswith(":"):
+                return f"{tokens[ndx + 1]}_{token[:-1]}"
```
Token with `:` is never last in reversed order as there is always visibility specifier. The only exception is `crate` that is handled separately

igorostrowskiq commented on scripts/coverage_checker.py:
```diff
@@ -0,0 +1,346 @@
+        for node in self.root.traverse_depth_first_pre_order():
+            children_to_remove = []
+            for ndx, child in enumerate(node.children):
+                if child.visibility != visibility.value:
```
Fixed

PiotrKorkus commented on scripts/coverage_checker/README.md:
```diff
@@ -0,0 +1,36 @@
+  * Install by `cargo install cargo-modules`
+
+## Example usage:
+`cd <inc_orchestrator_repo>`
+`cargo modules structure --package orchestration | sed 's/\x1b\[[0-9;]*m//g' > input_report.txt` - sed is required to remove colored output
+`python3 scripts/coverage_checker.py -r <older_orchestration_report> input_report.txt`
```
* It doesn't look good when rendered, replace it with single block of code ```.
* Path in commands shall be relative to README, so just `coverage_checker.py`
* Also please install https://marketplace.visualstudio.com/items?itemName=DavidAnson.vscode-markdownlint and solve issues

igorostrowskiq commented:
> Take a look into copilot comments. They aren't that bad and even caught a bug. I would suggest following structure:
> 
>     * create an extra directory `coverage_checker` in scripts and place script there
> 
>     * add bash script with command to generate input file
>       We will need some docstrings for sure.
>       Let's have a call, as when I try to run it I get quite messy tabs in the output.

1. Moved to directory `coverage_checker`
2. Added readme with commandline
3. Added docstrings

PiotrKorkus reviewed (commented):
Please add information about setting `tabs 4` for best output representation
